### ISSUE TITLE
Support for COLORTERM

### ISSUE THREAD

**Issue by moonglum** [tmux/tmux#4300]
### Issue description

I recently switched from foot to Ptyxis, a VTE-based terminal. This causes issues for me for vim running inside tmux, as vim only runs with 256 colors. I could resolve this issue with setting `terminal-features`, but I think this should work automatically instead.

From my understanding, tmux has [special handling for iTerm 2 and foot](https://github.com/tmux/tmux/blob/ae8f2208c98e3c2d6e3fe4cad2281dce8fd11f31/tty-features.c#L467-L503). But I think it would work for a lot more terminals if, instead, it checked if `$COLORTERM` is set to `truecolor`. [This is also described here](https://github.com/termstandard/colors?tab=readme-ov-file#truecolor-detection). What do you think?

Thanks for working on tmux, I'm using it for years and really appreciate it :heart:

### Required information

Please provide the following information:

* tmux version (`tmux -V`): `tmux 3.5a`
* Platform (`uname -sp`): `Linux unknown` (it is Arch)
* $TERM inside and outside of tmux (`echo $TERM`): `tmux-256color`
* Logs from tmux (`tmux kill-server; tmux -vv new`): Let me know if this makes sense in this case

**Comment by nicm** [comment]
I don't really think more environment variables are the solution to anything.

**Comment by moonglum** [comment]
Okay, what do you think would be the right solution? Using `*-direct` as `TERM`?

**Comment by egmontkob** [comment]
Let's link to the discussion that prompted this request: https://gitlab.gnome.org/chergert/ptyxis/-/issues/320.

`COLORTERM=truecolor` is a widely used practice, due to terminfo not having a usable way of denoting truecolor support. It's very far from ideal, but unfortunately this is the best the terminal emulation ecosystem has at the moment.

**Comment by nicm** [comment]
You can use `*-direct` or if you don't like the fact that it stops 256 colour `setaf`/`setab` working then could add `Tc` or `setrgbf`/`setrgbb`.

**Comment by nicm** [comment]
This will support COLORTERM but I'm not really sold on it TBH [tmux-colorterm.diff.txt](https://github.com/user-attachments/files/18176487/tmux-colorterm.diff.txt).

**Comment by nicm** [comment]
My concern is that there are perfectly good terminfo(5) capabilities to advertise RGB colour in the same way as every other terminal capability rather than adding a separate thing just for this...

**Comment by moonglum** [comment]
I definitely support having fewer special cases. Unfortunately, terminals just have a lot of historical baggage, and this is not always possible.

From my understanding, without the [special handling in tmux](https://github.com/tmux/tmux/blob/ae8f2208c98e3c2d6e3fe4cad2281dce8fd11f31/tty-features.c#L467-L503) other terminal emulators like iTerm2 would also not display RGB colors. The special handling for iTerm2’s RGB colors could therefore be removed from features.c and the patch you posted above would work for iTerm2 as well as the VTE-based ones (maybe that’s true for other ones in that special case block as well, but I know it is true for iTerm2 as I tested on a friend’s device that iTerm2 sets COLORTERM and introduces itself as xterm-256color like the Gnome ones do).

**Comment by nicm** [comment]
No, it couldn't be removed because the existing detection of iTerm2 always works over ssh and COLORTERM may not.

**Comment by moonglum** [comment]
As I tried to explain above, I don’t think there is an easy answer here. Terminals have so many layers of compatibilities, and hacks, that it is not easy to move forward. You need to decide if this is the right trade-off for your software. Feel free to close the ticket if you think it is the wrong decision for tmux. Thanks for taking the time considering it, and for tmux 🙇

**Comment by pedronavf** [comment]
That logic can be expressed by a configuration file:

```
# Set the RGB capability if the environment variable COLORTERM is truecolor or 24 bit
%if "#{==:#{COLORTERM},truecolor}"
  set -asF terminal-features "#{TERM}:RGB"
%elif "#{==:#{COLORTERM},24bit}"
  set -asF terminal-features "#{TERM}:RGB"
%endif
```

Some format variables, like `#{client_termname}` are not available at the time the configuration file is parsed (that's why I had to use the TERM variable in the code above), but if they were the code that's currently in tty_default_features could look like this:

```
set -s @tty_features_modern_xterm '256,RGB,bpaste,clipboard,mouse,strikethrough,title'

%if "#{==:#{client_termname},mintty}"
  set -asF terminal-features "mintty:#{@tty_features_modern_xterm},ccolour,cstyle,extkeys,margins,overline,usstyle"
%elif "#{==:#{client_termname},tmux}"
  set -asF terminal-features "tmux:#{@tty_features_modern_xterm},ccolour,cstyle,focus,overline,usstyle,hyperlinks"
<...>
%endif
```

Maybe tmux could have a configuration file loaded before the user one, making all this logic easily updateable and capturing these special use cases?

**Comment by PatTheMav** [comment]
> That logic can be expressed by a configuration file:
> 
> ```
> # Set the RGB capability if the environment variable COLORTERM is truecolor or 24 bit
> %if "#{==:#{COLORTERM},truecolor}"
>   set -asF terminal-features "#{TERM}:RGB"
> %elif "#{==:#{COLORTERM},24bit}"
>   set -asF terminal-features "#{TERM}:RGB"
> %endif
> ```

Wouldn't it be better to use `terminal-features` here, like so:

```
%if "#{==:#{COLORTERM},truecolor}"
    set -as terminal-features ",*-256color:RGB"
    [...]
%endif
```

This seems to reflect the reality of most systems, where Terminals only announce themselves as `xterm-256color` but set `COLORTERM` to additionally mark themselves as capable of displaying/understanding hexadecimal RGB colour values as well.

(See also [the terminfo rant](https://twoot.site/@bean/113056942625234032) about how `terminfo` causes more harm than good in 2024 and beyond).

Terminal.app in macOS 26 has gained 24-bit colour support and follows exactly the example I outlined above and I don't expect an updated terminfo entry for it and if it is updated, that update will only percolate slowly.

I've also noticed that at least on my Debian boxes there's a `xterm-direct256` entry which allows both direct colour mapping as well as access to 256 indexed colours (and thus will not break display of any colours apart from the 8 primary ANSI colours) and thus behaves exactly as `xterm-256color` with `COLORTERM` announced as `truecolor`.

**Comment by mark7** [comment]
> I've also noticed that at least on my Debian boxes there's a xterm-direct256 entry which allows both direct colour mapping as well as access to 256 indexed colours (and thus will not break display of any colours apart from the 8 primary ANSI 

If I understand aright, the basic problem with rolling out functionality via a new terminfo entry is that TERM must take on a new value.  Okay, fine, but that creates problems, as the TERM scheme has no fallback system if a given terminal type is not supported on the server because its terminfo database has not been updated.  A client has no way of knowing what terminal types a server supports (absent out-of-band manual user configuration), so it cannot provide for graceful fallback either.  This means that rolling out a new terminfo entry to support new functionality breaks functionality on servers that don't know that new terminal type until they also gain support.

So, say I have a scenario like this:

* Xterm (or some other virtual terminal package) gets new functionality.

* A new terminfo type is rolled out to support this functionality (say, "xterm-256color", as has happened in the past)
   
* Server 1 has a new terminfo, with an entry for "xterm-256color".  Server 2 does not.

Under that scenario, if a user SSHing from a client machine has TERM=xterm, then on Server 1, they lack 256 color support.  If a user SSHing from a client machine has TERM=xterm-256color, then on Server2, they have a not-understood terminfo entry (which as I recall from past use, tends to make software fall back to an assumption of a "dumb" terminal type).  As things stand, eventually this can be fixed when all systems get updated terminfo databases (though this may take some time or never happen with some embedded devices).  In the meantime, the user probably has to set TERM on a per-server basis to work around terminfo differences across servers.  That makes using "use a new terminfo type" a painful route to take.

It seems to me that the original sin here was not building any mechanism for fallback support into the TERM system.  That is, it would be better to have a scheme where one passed something like FALLBACK_TERM="xterm-256color:xterm" containing a list of desired TERM types in decreasing order of preference.  A server (which may or may not know about a new virtual terminal) could simply iterate down the list until it finds a type for which it has a terminfo entry, providing for graceful fallback, and a client --- which must know about the new virtual terminal if it is to be installed on the client --- can be responsible for having its terminal provide the fallback list.

Now that we're on 24-bit-color, the fallback list would look like FALLBACK_TERM="xterm-direct256:xterm-256color:xterm".  TERM could remain in place for older systems that don't understand FALLBACK_TERM, and simply list the first item in the FALLBACK_TERM list. 

That seems to me like a more-comprehensive fix for the basic problem that it's hard to roll out new functionality via new terminfo entries.

I realize that this is a bit out-of-scope for this issue, but since the discussion here is over adding additional environment variables to communicate terminal capabilities, I have to say that it seems like if new environment variables are to be introduced, it seems like it'd be better to create a comprehensive fix rather than a one-off fix that will provide information about support for one particular feature (24-bit color).  I doubt that 24-bit color will be the last feature that is added to virtual terminals.

**Comment by ThomasDickey** [comment]
None of the preceding is original thinking.  See for example [Debian [#357742](mailto:357742@bugs.debian.org)](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=357742).  A patch (demonstrating how to work around the defects in the idea) might be a basis for discussion.

**Comment by nicm** [comment]
I've applied this change now, I don't like it much but I can't really see it doing any harm. Thanks!
